### PR TITLE
Tests run on Docker

### PR DESCRIPTION
**Pull request by LysandreJik**
This PR updates the GPU-based tests to run on docker images. This:
- Simplifies maintenance
- Horizontal scaling is made much simpler
- Environment can be managed with a single line change (the docker image) hello pytorch 1.3-1.7 tests!
- Setups a notification service to get alerted when scheduled tests fail.

Co-authored-by: Morgan <funtowiczmo@gmail.com>

### CODE REVIEW

File: .github/workflows/self-push.yml [hunk 1]
```diff
@@ -10,73 +10,43 @@ on:
...
-  # pull_request:
   repository_dispatch:
 
-
 jobs:
   run_tests_torch_gpu:
-    runs-on: [self-hosted, gpu, single-gpu]
+    runs-on: [self-hosted, docker-gpu, single-gpu]
+    container:
+      image: pytorch/pytorch:1.8.0-cuda11.1-cudnn8-runtime
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:19]: Based on Pytorch's image

File: .github/workflows/self-push.yml [hunk 2]
```diff
@@ -10,73 +10,43 @@ on:
...
+      options: --gpus all --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:20]: Shared cache across machines

File: .github/workflows/self-push.yml [hunk 3]
```diff
@@ -89,68 +59,38 @@ jobs:
...
-          python --version
-          pip --version
-
-      - name: Current dir
-        run: pwd
-
-      - run: nvidia-smi
-
-      - name: Kill any run-away pytest processes
-        run: (pkill -f tests; pkill -f examples) || echo "no zombies"
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:109]: @stas00 those are not needed anymore since the docker container is discarded at the end of the the suite

File: .github/workflows/self-push.yml [hunk 4]
```diff
@@ -89,68 +59,38 @@ jobs:
...
-          pip --version
+          nvidia-smi
 
       - name: Install dependencies
         run: |
-          source .env/bin/activate
           pip install --upgrade pip
-          pip install .[tf,sklearn,testing,onnxruntime,sentencepiece]
-          pip install git+https://github.com/huggingface/datasets
+          pip install .[sklearn,testing,onnxruntime,sentencepiece]
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:78]: No need to install TF as it's already installed in the image

File: .github/workflows/self-push.yml [hunk 5]
```diff
@@ -89,68 +59,38 @@ jobs:
           name: run_all_tests_torch_gpu_test_reports
           path: reports
 
-
   run_tests_tf_gpu:
-    runs-on: [self-hosted, gpu, single-gpu]
+    runs-on: [self-hosted, docker-gpu, single-gpu]
+    container:
+      image: tensorflow/tensorflow:2.4.1-gpu
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:65]: TensorFlow's docker image

File: .github/workflows/self-push.yml [hunk 6]
```diff
@@ -10,73 +10,43 @@ on:
...
-#          source .env/bin/activate
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/pt-encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/standalone.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/tf-encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-
       - name: Run all non-slow tests on GPU
         env:
-          OMP_NUM_THREADS: 1
+          OMP_NUM_THREADS: 8
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:45]: This will be automated later, but we have 16 vCPUs so let's take advantage of them. (and two workers in Pytest, hence why we put 8 and not 16)

File: .github/workflows/self-push.yml [hunk 7]
```diff
@@ -10,73 +10,43 @@ on:
...
 
       - name: Install dependencies
         run: |
-          source .env/bin/activate
-          sudo apt-get -y update && sudo apt-get install -y libsndfile1-dev
+          apt -y update && apt install -y libsndfile1-dev
           pip install --upgrade pip
-          pip install .[torch,sklearn,testing,onnxruntime,sentencepiece,speech]
-          pip install git+https://github.com/huggingface/datasets
+          pip install .[sklearn,testing,onnxruntime,sentencepiece,speech]
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:33]: No need to specify torch as it's already installed in the image

File: setup.py [hunk 8]
```diff
@@ -115,6 +115,7 @@
     "psutil",
     "pydantic",
     "pytest",
+    "pytest-sugar",
```
Comment (LysandreJik) [inline setup.py:118]: This addition makes it much easier to monitor how the slow tests are going. Beforehand, we had no way of knowing at which % the slow tests were, now if one opens the run it will show how much of the suite has ran.

File: setup.py [hunk 9]
```diff
@@ -225,14 +226,15 @@ def run(self):
 
 extras["tokenizers"] = deps_list("tokenizers")
 extras["onnxruntime"] = deps_list("onnxruntime", "onnxruntime-tools")
+extras["onnx"] = deps_list("onnxconverter-common", "keras2onnx") + extras["onnxruntime"]
```
Comment (LysandreJik) [inline setup.py:229]: So that all of these can be installed in the TF suite.

File: tests/test_modeling_tf_common.py [hunk 10]
```diff
@@ -129,7 +129,7 @@ def test_save_load(self):
 
                 self.assert_outputs_same(after_outputs, outputs)
 
-    @slow
+    @unittest.skip("Skipping this as too slow")
```
Comment (LysandreJik) [inline tests/test_modeling_tf_common.py:132]: Will reinstate those later, but they're way too slow for now.

**Review by sgugger — APPROVED** [review]
Awesome! Thank you so much for diving into this!

File: .github/workflows/self-push.yml [hunk 11]
```diff
@@ -10,73 +10,43 @@ on:
...
 
       - name: Are GPUs recognized by our DL frameworks
+        env:
+          CUDA_VISIBLE_DEVICES: 0
         run: |
-          source .env/bin/activate
           python -c "import torch; print('Cuda available:', torch.cuda.is_available())"
+          python -c "import torch; print('Cuda version:', torch.version.cuda)"
```
Comment (mfuntowicz) [inline .github/workflows/self-push.yml:40]: Do we want to print cudnn version also ? 

```bash
python -c "import torch; print('CuDNN version:', torch.backends.cudnn.version)"
```

File: .github/workflows/self-push.yml (hunk 6 above)
Comment (mfuntowicz) [reply .github/workflows/self-push.yml:45]: Let's add  `MKL_NUM_THREADS: 8`  too here, to be sure we are not oversubscribing cores to multiple workers.

File: .github/workflows/self-push.yml [hunk 12]
```diff
@@ -89,68 +59,38 @@ jobs:
...
-          source .env/bin/activate
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/pt-encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/standalone.json --path=templates/adding_a_new_model
-#          transformers-cli add-new-model --testing --testing_file=templates/adding_a_new_model/tests/tf-encoder-bert-tokenizer.json --path=templates/adding_a_new_model
-
       - name: Run all non-slow tests on GPU
         env:
-          OMP_NUM_THREADS: 1
+          OMP_NUM_THREADS: 8
```
Comment (mfuntowicz) [inline .github/workflows/self-push.yml:89]: Same here, i would add: 

```yaml
MKL_NUM_THREADS: 8
TF_NUM_INTEROP_THREADS: 1
```

`TF_NUM_INTEROP_THREADS` controls the number of threads involved in everything not compute bound, which essentially reflects graph control flow operations.

File: .github/workflows/self-push.yml [hunk 13]
```diff
@@ -89,68 +59,38 @@ jobs:
           name: run_all_tests_torch_gpu_test_reports
           path: reports
 
-
   run_tests_tf_gpu:
-    runs-on: [self-hosted, gpu, single-gpu]
+    runs-on: [self-hosted, docker-gpu, single-gpu]
+    container:
+      image: tensorflow/tensorflow:2.4.1-gpu
+      options: --gpus all --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
```
Comment (mfuntowicz) [inline .github/workflows/self-push.yml:66]: Do you need `--gpus all` if you want to run only on a single GPU? 

You can do `--gpus  0` and then remove relying on `CUDA_VISIBLE_DEVICES: 0`

File: .github/workflows/self-push.yml (hunk 2 above)
Comment (mfuntowicz) [inline .github/workflows/self-push.yml:20]: Do you need --gpus all if you want to run only on a single GPU?

You can do --gpus 0 and then remove relying on CUDA_VISIBLE_DEVICES: 0

**Review by mfuntowicz — APPROVED** [review]

File: setup.py (hunk 8 above)
Comment (stas00) [reply setup.py:118]: you might want to add this one too:
https://huggingface.co/transformers/testing.html#instantly-shows-failed-tests

I used to use `pytest-sugar` for a while but then I remember there was some problem that I removed it but I can't recall now which.

File: .github/workflows/self-push.yml (hunk 13 above)
Comment (stas00) [reply .github/workflows/self-push.yml:66]: We must make it 1 gpu or as you said set `CUDA_VISIBLE_DEVICES: 0` because some tests will be skipped if it finds multiple-gpus.

File: tests/test_modeling_tf_common.py (hunk 10 above)
Comment (stas00) [reply tests/test_modeling_tf_common.py:132]: making a new - `@tooslow` decorator?

**Review by stas00 — APPROVED** [review]
Looks wonderful!

My only concern is that we are no longer testing installing dependencies.

Perhaps there should be one old-style scheduled job that tests a full install as it's prior to this PR. e.g. this could be in addition to the infrequent tests of older pytorch versions we discussed elsewhere.

and we need a doc on how to update the docker image when we need to update the env.

File: .github/workflows/self-push.yml (hunk 13 above)
Comment (LysandreJik) [reply .github/workflows/self-push.yml:66]: Yes, it's currently using `CUDA_VISIBLE_DEVICES: 0` but that's just a check as these runs only run on single-GPU machines anyway. I can add `--gpus 0` instead of `CUDA_VISIBLE_DEVICES: 0`, but again these two should be unnecessary as the tag `single-gpu` prevents these from being run on multi-GPU machines.

Comment (stas00) [reply .github/workflows/self-push.yml]: If it's not too much work, my vote is to make it explicit, since the snippets later get copied and when reading several of these in parallel it's easier to quickly grasp that this is 1-gpu recipe.

**Comment by LysandreJik** [comment]
@stas00 I hear you regarding the installation of dependencies. We are installing every dependency except the main one (PT/TF) on the GPU machine, so I would say most errors are caught.
Furthermore, we are installing these exact dependencies on the CircleCI runs, so I believe we are testing that already on every commit.

If there is a scenario I am missing, please let me know and I will do my best to adjust.

All other comments have been adressed in the last three commits.

**Comment by stas00** [comment]
Ah, ok, for some reason I thought that if we are using a docker image then we can skip wasting time and resources on installing the same things million times a day, and in that case only an occasional test that installs from scratch will be needed. But perhaps this further speed up can be done in some future iteration.

Nothing more from my side.

Thank you, @LysandreJik
